### PR TITLE
Handle duplicate column names in select_dtypes and get_dummies

### LINKED ISSUE #20848: select_dtypes and get_dummies break on duplicate dolumns
Functions `select_dtypes` and `get_dummies` have strange and incorrect behavior on duplicate column names. Shown below:

```python
In [6]: df
Out[6]: 
  col1 col1
0    1    a
1    2    b

In [7]: df.select_dtypes(include=['int'])
Out[7]: 
Empty DataFrame
Columns: []
Index: [0, 1]

In [8]: pd.get_dummies(df)
Out[8]: 
   col1_('c', 'o', 'l', '1')  col1_('c', 'o', 'l', '1')
0                          1                          1
1                          1                          1
```

### PR DESCRIPTION
**Pull request by kunalgosar**
Functions `select_dtypes` and `get_dummies` previously failed on DataFrames with duplicate column names and had strange behavior as shown below. This PR fixes that strange behavior.

Previous behavior:

```python
In [6]: df
Out[6]: 
  col1 col1
0    1    a
1    2    b

In [7]: df.select_dtypes(include=['int'])
Out[7]: 
Empty DataFrame
Columns: []
Index: [0, 1]

In [8]: pd.get_dummies(df)
Out[8]: 
   col1_('c', 'o', 'l', '1')  col1_('c', 'o', 'l', '1')
0                          1                          1
1                          1                          1
```

New behavior:

```python
In [6]: df
Out[6]: 
  col1 col1
0    1    a
1    2    b

In [7]: df.select_dtypes(include=['int'])
Out[7]: 
   col1
0     1
1     2

In [8]: pd.get_dummies(df)
Out[8]: 
   col1  col1_a  col1_b
0     1       1       0
1     2       0       1
```

- [x] closes #20848
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
does this have an open issue?

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -826,45 +826,49 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
...
+            prefix_sep = [prefix_sep[col] for col in columns_to_encode.columns]
 
-        if set(columns_to_encode) == set(data.columns):
+        if columns_to_encode.shape == data.shape:
             with_dummies = []
+        elif columns is not None:
+            with_dummies = [data.drop(columns, axis=1)]
         else:
-            with_dummies = [data.drop(columns_to_encode, axis=1)]
+            with_dummies = [data.select_dtypes(exclude=['object', 'category'])]
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: why did you add this?

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -826,45 +826,49 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
...
         # validate prefixes and separator to avoid silently dropping cols
         def check_len(item, name):
             len_msg = ("Length of '{name}' ({len_item}) did not match the "
                        "length of the columns being encoded ({len_enc}).")
 
             if is_list_like(item):
-                if not len(item) == len(columns_to_encode):
-                    len_msg = len_msg.format(name=name, len_item=len(item),
-                                             len_enc=len(columns_to_encode))
+                if not len(item) == columns_to_encode.shape[1]:
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: are you sure this is correct?

File: pandas/core/reshape/reshape.py [hunk 3]
```diff
@@ -826,45 +826,49 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
 
         if columns is None:
             columns_to_encode = data.select_dtypes(
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: this is now not columns, so rename to something else

File: pandas/tests/frame/test_dtypes.py [hunk 4]
```diff
@@ -287,6 +287,21 @@ def test_select_dtypes_include_exclude_mixed_scalars_lists(self):
         ei = df[['b', 'c', 'f', 'k']]
         assert_frame_equal(ri, ei)
 
+    def test_select_dtypes_duplicate_columns(self):
+        df = DataFrame({'a': list('abc'),
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: add the PR number as a comment

File: pandas/tests/frame/test_dtypes.py [hunk 5]
```diff
@@ -287,6 +287,21 @@ def test_select_dtypes_include_exclude_mixed_scalars_lists(self):
...
+                        'b': list(range(1, 4)),
+                        'c': np.arange(3, 6).astype('u1'),
+                        'd': np.arange(4.0, 7.0, dtype='float64'),
+                        'e': [True, False, True],
+                        'f': pd.date_range('now', periods=3).values})
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: this construction is actually not guaranteed under < py3.6, so use an OrderdDict

File: pandas/tests/frame/test_dtypes.py [hunk 6]
```diff
@@ -287,6 +287,21 @@ def test_select_dtypes_include_exclude_mixed_scalars_lists(self):
...
+        df.columns = ['a', 'a', 'b', 'b', 'b', 'c']
+
+        e = DataFrame({'a': list(range(1, 4)),
```
Comment (jreback) [inline pandas/tests/frame/test_dtypes.py]: e -> expected, r -> result

File: pandas/tests/reshape/test_reshape.py [hunk 7]
```diff
@@ -465,6 +465,20 @@ def test_get_dummies_dont_sparsify_all_columns(self, sparse):
 
         tm.assert_frame_equal(df[['GDP']], df2)
 
+    def test_get_dummies_duplicate_columns(self, df):
+        df.columns = ["A", "A", "A"]
```
Comment (jreback) [inline pandas/tests/reshape/test_reshape.py]: issue number

**Review by jreback — CHANGES_REQUESTED** [review]
looks ok, but need to have a look after you update comments

File: pandas/core/reshape/reshape.py [hunk 8]
```diff
@@ -825,46 +825,50 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
         # determine columns being encoded
 
         if columns is None:
-            columns_to_encode = data.select_dtypes(
-                include=['object', 'category']).columns
+            data_to_encode = data.select_dtypes(
+                include=['object', 'category'])
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: can you define these at the top of the function as they are used in 2 places

File: pandas/core/reshape/reshape.py [hunk 9]
```diff
@@ -825,46 +825,50 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
...
         else:
-            with_dummies = [data.drop(columns_to_encode, axis=1)]
+            with_dummies = [data.select_dtypes(exclude=['object', 'category'])]
 
-        for (col, pre, sep) in zip(columns_to_encode, prefix, prefix_sep):
+        for (col, pre, sep) in zip(data_to_encode.iteritems(), prefix,
+                                   prefix_sep):
 
-            dummy = _get_dummies_1d(data[col], prefix=pre, prefix_sep=sep,
+            dummy = _get_dummies_1d(col[1], prefix=pre, prefix_sep=sep,
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: can you put a comment why are using col[1] here (IOW what is that)

File: pandas/core/reshape/reshape.py [hunk 10]
```diff
@@ -825,46 +825,50 @@ def get_dummies(data, prefix=None, prefix_sep='_', dummy_na=False,
...
+            prefix = data_to_encode.columns
 
         # validate separators
         if isinstance(prefix_sep, compat.string_types):
             prefix_sep = cycle([prefix_sep])
         elif isinstance(prefix_sep, dict):
-            prefix_sep = [prefix_sep[col] for col in columns_to_encode]
+            prefix_sep = [prefix_sep[col] for col in data_to_encode.columns]
 
-        if set(columns_to_encode) == set(data.columns):
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: can you add comments for these cases (IOW what they are for)

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. some doc comments

**Review by jreback — CHANGES_REQUESTED** [review]
need a whatsnew note as well, you can put in 0.23.0 (if you can update in next few days)

**Comment by kunalgosar** [comment]
Thanks @jreback. I resolved comments and added the whatsnew note.

**Comment by kunalgosar** [comment]
Rebased on master so checks pass now @jreback.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @kunalgosar nice patch!

**Pull request merged**